### ISSUE TITLE
Stalled and unkillable COPY FROM

### ISSUE THREAD

**Issue by benjaminrigaud** [crate/crate#3317]
Hello,

- Crate 0.54.5

I launched a COPY FROM with one set of json files (generated by a COPY TO) of ~200M documents
```
COPY events_2015_10 FROM '/tmp/backup_crate/events_2015_10/*' WITH(compression='gzip')
```
1) It imported about ~100M documents (in few hours) then stopped: the load decreased and no documents were added.
2) The query is still in sys.jobs
3) A kill id or kill all does nothing

It looks like it did one shard then freezed.

Ben

**Comment by benjaminrigaud** [comment]
I tried again by splitting my COPY FROM shard by shard: it worked quite well for a time then stalled again.

**Comment by mfussenegger** [comment]
Would it be possible to provide a thread-dump of the nodes once the query becomes frozen?

**Comment by benjaminrigaud** [comment]
Too late:/ I just restarted the cluster.
How should I do that next time?

I have this log on one node:
```
[2016-03-03 09:48:39,244][INFO ][node                     ] [H.E.R.B.I.E.] stopping ...
[2016-03-03 09:48:39,285][WARN ][netty.channel.DefaultChannelPipeline] An exception was thrown by an exception handler.
java.util.concurrent.RejectedExecutionException: Worker has already been shutdown
        at org.jboss.netty.channel.socket.nio.AbstractNioSelector.registerTask(AbstractNioSelector.java:120)
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.executeInIoThread(AbstractNioWorker.java:72)
        at org.jboss.netty.channel.socket.nio.NioWorker.executeInIoThread(NioWorker.java:36)
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.executeInIoThread(AbstractNioWorker.java:56)
        at org.jboss.netty.channel.socket.nio.NioWorker.executeInIoThread(NioWorker.java:36)
        at org.jboss.netty.channel.socket.nio.AbstractNioChannelSink.execute(AbstractNioChannelSink.java:34)
        at org.jboss.netty.channel.DefaultChannelPipeline.execute(DefaultChannelPipeline.java:636)
        at org.jboss.netty.channel.Channels.fireExceptionCaughtLater(Channels.java:496)
        at org.jboss.netty.channel.AbstractChannelSink.exceptionCaught(AbstractChannelSink.java:46)
        at org.jboss.netty.channel.DefaultChannelPipeline.notifyHandlerException(DefaultChannelPipeline.java:658)
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:781)
        at org.jboss.netty.channel.Channels.write(Channels.java:725)
        at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.doEncode(OneToOneEncoder.java:71)
        at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.handleDownstream(OneToOneEncoder.java:59)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:591)
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:784)
        at org.elasticsearch.http.netty.pipelining.HttpPipeliningHandler.handleDownstream(HttpPipeliningHandler.java:87)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:591)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:582)
        at io.crate.http.netty.NettyHttpChannel.sendResponse(NettyHttpChannel.java:199)
        at io.crate.rest.action.RestSQLAction$SQLResponseListener.onFailure(RestSQLAction.java:147)
        at io.crate.action.sql.TransportBaseSQLAction.sendResponse(TransportBaseSQLAction.java:212)
        at io.crate.action.sql.TransportBaseSQLAction.access$100(TransportBaseSQLAction.java:80)
        at io.crate.action.sql.TransportBaseSQLAction$2.onFailure(TransportBaseSQLAction.java:270)
        at com.google.common.util.concurrent.Futures$6.run(Futures.java:1310)
        at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:457)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:202)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setExceptionAndMaybeLog(Futures.java:1704)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1755)
        at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1608)
        at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1686)
        at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:457)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:202)
        at com.google.common.util.concurrent.SettableFuture.setException(SettableFuture.java:68)
        at io.crate.operation.QueryResultRowDownstream.fail(QueryResultRowDownstream.java:64)
        at io.crate.operation.projectors.AggregationPipe.fail(AggregationPipe.java:76)
        at io.crate.operation.merge.IteratorPageDownstream.fail(IteratorPageDownstream.java:204)
        at io.crate.operation.merge.IteratorPageDownstream$3.onFailure(IteratorPageDownstream.java:145)
        at io.crate.core.MultiFutureCallback.countdown(MultiFutureCallback.java:67)
        at io.crate.core.MultiFutureCallback.onFailure(MultiFutureCallback.java:58)
        at com.google.common.util.concurrent.Futures$6.run(Futures.java:1310)
        at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:457)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:202)
        at com.google.common.util.concurrent.SettableFuture.setException(SettableFuture.java:68)
        at io.crate.jobs.PageDownstreamContext.failure(PageDownstreamContext.java:161)
        at io.crate.executor.transport.ExecutionPhasesTask$DirectResponseListener.onFailure(ExecutionPhasesTask.java:275)
        at io.crate.executor.transport.DefaultTransportResponseHandler.handleException(DefaultTransportResponseHandler.java:62)
        at org.elasticsearch.transport.TransportService$2.run(TransportService.java:178)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
[2016-03-03 09:48:39,453][INFO ][node                     ] [H.E.R.B.I.E.] stopped
[2016-03-03 09:48:39,454][INFO ][node                     ] [H.E.R.B.I.E.] closing ...
[2016-03-03 09:48:39,468][INFO ][node                     ] [H.E.R.B.I.E.] closed
```

**Comment by mfussenegger** [comment]
You can send a SIGQUIT signal to the java process - that will cause it do dump the thread dump to STDOUT. Or you can use jstack. See https://helpx.adobe.com/experience-manager/kb/TakeThreadDump.html

**Comment by benjaminrigaud** [comment]
jstack doesn't work:

`Unable to open socket file: target process not responding or HotSpot VM not loaded
`
Do you I need to run Crate with the java from the jdk instead of the jre?

I run Crate on Ubuntu as a service, is the STDOUT logged somewhere?

**Comment by mfussenegger** [comment]
Where the stdout of crate ends up depends on your distribution and how you've started crate.
As you're on Ubuntu 14.04 which by default uses upstart the output should end up in `/var/log/upstart/crate.log`

**Comment by benjaminrigaud** [comment]
thank you! Am such a noob with upstart.
I did a dump on one node. I'll do all the nodes as soon as I reproduce the issue (I got it 2 times today).


[Uploading dump.stack.crate.txt…]()

**Comment by mfussenegger** [comment]
Okay, looking forward to the other stack traces. Just make sure that you take the dumps after the query has gotten stuck (and before you try to kill it)

**Comment by mfussenegger** [comment]
We've encountered the same issue on one of our clusters, so we're able to investigate this.

**Comment by benjaminrigaud** [comment]
great news! Tell me if you need anything else.

Have a nice week-end.

**Comment by mfussenegger** [comment]
Unfortunately we haven't been able to reproduce this issue again after friday.

But after some investigation I've found two things that could cause a copy from statement to appear stuck although it isn't.

 - On copy from s3 I've seen cases where the download speed from s3 gets really slow. A tool like [iftop](https://en.wikipedia.org/wiki/Iftop) can be used to verify if this is the case.
 - The nodes reading the files are being slowed down / paused if the nodes doing the indexing are being overloaded. The resume/speed up after the indexing nodes recover can be a bit sluggish.


In your case you imported from the filesystem right? How long did you wait for the query to complete after the load dropped?

**Comment by benjaminrigaud** [comment]
- Am importing from the filesystem yes.
- Few hours :) The load have been really low for a really long time before I tried to kill the query.

**Comment by chaudum** [comment]
one more question: how many files does the set of json files contain? are they all on one single node, or distributed across all nodes?

**Comment by benjaminrigaud** [comment]
They were generated by a COPY TO: about 2 files per table (one per shard) on 2 different nodes.

**Comment by chaudum** [comment]
The "unkillable job" may be fixed with https://github.com/crate/crate/commit/ba03028d6d1f460b6323d262b6429f19c932470f

**Comment by joemoe** [comment]
hi @benjaminrigaud, just to keep you up to date on this one. something very similar happened by another user and we are close to a fix.

**Comment by benjaminrigaud** [comment]
great news.

**Comment by joemoe** [comment]
... and another update. we were finally able to narrow it down and we already got a fix in master. so if you'd be able to try it out with a nightly this would help us a lot.
i'll tell you whenever it is in a release.

**Comment by benjaminrigaud** [comment]
I won't be able to test soon: I only have one production cluster (and I don't want to break it yet :p)
Thank you.

**Comment by joemoe** [comment]
@benjaminrigaud we thought we had a fix in 0.54.9, but now found out that something similar might happen. did you already give 0.54.9 a try?

**Comment by benjaminrigaud** [comment]
Hello, I am using 0.54.9 but am not doing big COPY FROM at this time (20 - 20000 docs currently).
I really need to setup a staging cluster to try this issue: I won't be able to do that soon.

**Comment by joemoe** [comment]
ok, we also think we now found the real cause and it will be in 0.54.10 and 0.55.0.

**Comment by joemoe** [comment]
hi @benjaminrigaud, we released the testing release of 0.55 last week. would be cool if you could give it a try.

**Comment by benjaminrigaud** [comment]
It should be possible locally : I'll setup a docker cluster on my dev laptop.

**Comment by joemoe** [comment]
thanks.

**Comment by joemoe** [comment]
@benjaminrigaud i'm closing this issue now as this is solved with 0.54.9 and 0.55.0 for other users. if you are able to reproduce it with this versions please open it again.
